sample-tracking-status-overview-domain/src/test/groovy/life/qbic/business/project/subscribe/SubscribeProjectSpec.groovy

```diff
@@ -0,0 +1,99 @@
+
+import life.qbic.business.DataSourceException
+import spock.lang.Specification
+
+/**
+ * <b>Specification for the SubscribeProject use case</b>
+ *
+ * @since 1.0.0
+ */
+class SubscribeProjectSpec extends Specification {
+    final String validFirstName = "firstname"
+    final String validLastName = "lastname"
+    final String validEmail = "email@addre.ss"
+    final String validProjectCode = "QABCD"
+
+    SubscribeProjectOutput output = Mock()
+    SubscriptionDataSource subscriptionDataSource = Mock()
+    SubscribeProject subscribeProject = new SubscribeProject(subscriptionDataSource, output)
+
+    def "Subscribe fails for invalid first name: #invalidFirstName"() {
+        when:
+        subscribeProject.subscribe(invalidFirstName, validLastName, validEmail, validProjectCode)
+        then:
+        thrown(IllegalArgumentException)
+        where:
+        invalidFirstName << [null, ""]
+    }
+
+    def "Subscribe fails for invalid last name: #invalidLastName"() {
+        when:
+        subscribeProject.subscribe(validFirstName, invalidLastName, validEmail, validProjectCode)
+        then:
+        thrown(IllegalArgumentException)
+        where:
+        invalidLastName << [null, ""]
+    }
+
+    def "Subscribe fails for invalid email address: #invalidEmail"() {
+        when:
+        subscribeProject.subscribe(validFirstName, validLastName, invalidEmail, validProjectCode)
+        then:
+        thrown(IllegalArgumentException)
+        where:
+        invalidEmail << [null, ""]
+    }
+
+    def "Subscribe fails for invalid project code: #invalidProjectCode"() {
+        when:
+        subscribeProject.subscribe(validFirstName, validLastName, validEmail, invalidProjectCode)
+        then:
+        thrown(IllegalArgumentException)
+        where:
+        invalidProjectCode << [null, "", "1234", "ZBCA"]
+    }
+
+    def "Subscribe does not throw an IllegalArgumentException for valid arguments"() {
+        when:
+        subscribeProject.subscribe(validFirstName, validLastName, validEmail, validProjectCode)
+        then:
+        notThrown(IllegalArgumentException)
+    }
+
+    def "Subscribe informs output of success if no exception is thrown"() {
+        when:
+        subscribeProject.subscribe(validFirstName, validLastName, validEmail, validProjectCode)
+        then:
+        1 * output.subscriptionAdded(_)
+        0 * output.subscriptionFailed(_, _, _, _)
+    }
+
+    def "Subscribe informs output in case of data source failure"() {
+        given:
+        subscriptionDataSource = Stub()
+        subscriptionDataSource.subscribeToProject(_ as Subscriber,
+                _ as String) >> { throw new DataSourceException("Some exception.") }
+        subscribeProject = new SubscribeProject(subscriptionDataSource, output)
+        when:
+        subscribeProject.subscribe(validFirstName, validLastName, validEmail, validProjectCode)
+        then:
+        1 * output.subscriptionFailed(_, _, _, _)
```

Comment:
```suggestion
        1 * output.subscriptionFailed(validFirstName, validLastName, validEmail, validProjectCode)
```

Reply:
I am not sure whether this is what we want. In this case, the arguments are compared with the identity function ( `Object.is()`) and this fails for copied values. Having a check here on equality could be beneficial 👍

Reply:
You could just try if it works. At least it worked for my tests. You know the call arguments so for me it would only make sense to use them also for the output method call :)